### PR TITLE
PHPLIB-1137: Reject PackedArrays when expecting documents

### PR DESCRIPTION
**Pull request by alcaeus**
PHPLIB-1137

This PR adds a new `is_document` function that is used to validate input options. While we accept `array|object` for documents, a `PackedArray` instance is one of those objects where we are absolutely sure that it is not a document.

This PR also includes a commit to refactor most data providers that provide options to operation classes. This was done to give the individual data sets a name, making it easier to spot failing tests. Due to the large amount of unrelated changes in that first commit, I suggest reviewing the commits separately.

### CODE REVIEW

File: src/functions.php
Comment (jmikola) [inline src/functions.php:197]: Noting that we don't bother checking if array is a list, since APIs that take arrays as documents generally use an object cast to ensure proper BSON encoding as a document type.

File: tests/Operation/BulkWriteTest.php
Comment (jmikola) [inline tests/Operation/BulkWriteTest.php:60]: I don't recall why we accepted a space in the original pattern.

File: tests/Operation/BulkWriteTest.php
Comment (jmikola) [inline tests/Operation/BulkWriteTest.php:195]: I noted the `replaceOne` validation has three possible exceptions in order:

 * not a document
 * first key is an operator
 * is a pipeline (with `allowEmpty=true`)

Is the "update pipeline" still reachable because `is_document()` simply returns `true` for PHP arrays? What's the data set that triggers "not a document" to be reached here?

File: tests/Operation/BulkWriteTest.php
Comment (jmikola) [inline tests/Operation/BulkWriteTest.php:270]: I notice there's still an "array or object" check for `updateMany` and `updateOne`. Does it make sense to combine that with the `! is_first_key_operator && ! is_pipeline` check that appears later?

```
is_document && ! is_first_key_operator && ! is_pipeline
```

Or might that exclude PackedArray? Aggregation operations require an actual array for their pipelines, so I think it's just update and findAndModify ops that might subtlety allow a PackedArray because they currently accept `array|object`.

File: tests/Operation/BulkWriteTest.php
Comment (jmikola) [inline tests/Operation/BulkWriteTest.php:363]: Relates to my earlier question above about `updateMany` and `updateOne`.

File: tests/Operation/WatchTest.php
Comment (jmikola) [inline tests/Operation/WatchTest.php:53]: Good catch.

File: tests/TestCase.php
Comment (jmikola) [inline tests/TestCase.php:143]: Is this only used for Operation classes? If so, you can consider moving it to `MongoDB\Tests\Operation\TestCase`, which is where I put other providers.

File: tests/Collection/CollectionFunctionalTest.php
Comment (jmikola) [inline tests/Collection/CollectionFunctionalTest.php:64]: Why does this have a `void` return type? It looks like you're discarding the return value from `createOptionDataProvider()`.

File: tests/Command/ListCollectionsTest.php
Comment (jmikola) [inline tests/Command/ListCollectionsTest.php:18]: Likewise here. Why `void`?

File: tests/TestCase.php
Comment (jmikola) [inline tests/TestCase.php:177]: Does this potentially break if a data provider includes two values with the same type? I suppose PHPUnit would raise an error for that case, so we can leave as-is and address it if it ever becomes an issue down the line.

File: tests/TestCase.php
Comment (jmikola) [inline tests/TestCase.php:169]: This comment doesn't add much value. Is copilot the new copypasta?

Comment (alcaeus) [reply tests/TestCase.php:169]: The method was more complex originally (it involved a couple of closures), so I decided to add some comments. Then I removed the closure and made this easier but didn't remove the comment.

The only Copilot generated code in this PR is the repetitive `'<option>' => $this->getSomeValues(),` lines 😅

File: tests/Operation/BulkWriteTest.php
Comment (alcaeus) [reply tests/Operation/BulkWriteTest.php:195]: "Update pipeline" is indeed still reachable because of PHP arrays. Since some of the update pipelines are provided as `PackedArray` instances, these will trigger the "not a document" check before getting to the pipeline checks later on.

File: tests/Operation/BulkWriteTest.php
Comment (alcaeus) [reply tests/Operation/BulkWriteTest.php:270]: This indeed allows update pipelines to be passed as `PackedArray` instances. However, giving this some thought, the following options are valid:
- document with update operator as first key
- update pipeline (either as `PackedArray`, array, or `Serializable` that serialises to an array

With that in mind, we can refactor this to throw on this condition:
```
(! is_document || ! is_first_key_operator) && ! is_pipeline
```

I've made the change as the expected type (`array|object`) is less useful than the info that the operation requires an update document or update pipeline.

File: tests/Collection/CollectionFunctionalTest.php
Comment (alcaeus) [reply tests/Collection/CollectionFunctionalTest.php:64]: Good catch - not sure how that slipped through the cracks. I've updated this and the other occurrence and double-checked that I didn't make the same mistake more often.

File: tests/TestCase.php
Comment (alcaeus) [reply tests/TestCase.php:177]: PHPUnit wouldn't raise for that case, but we'd silently ignore tests. Since that is extremely hard to detect and can cause significant problems, I've changed this to use `<option>_<key>_<type>` as data set name for numeric keys, and `<option>_<key>` for string keys. That should remove all possible duplicates.

File: tests/TestCase.php
Comment (alcaeus) [reply tests/TestCase.php:143]: Ah, good point. I've moved this and `getInvalidUpdateValues` for consistency.

This also provides a nice refactoring opportunity down the line once we make all our data providers static (which is required in later PHPUnit versions): we can offload these data providers to separate classes or traits and import them as necessary.

File: tests/Operation/BulkWriteTest.php
Comment (alcaeus) [reply tests/Operation/BulkWriteTest.php:60]: "I'm sure it seemed like a good idea at the time" 😅

File: tests/Model/ChangeStreamIteratorTest.php
Comment (alcaeus) [reply tests/Model/ChangeStreamIteratorTest.php:70]: Fixed psalm errors in the respective commits.

File: src/functions.php
Comment (alcaeus) [reply src/functions.php:196]: Missed that, along with the parameter type hint. Added both.

File: src/Command/ListCollections.php
Comment (GromNaN) [inline src/Command/ListCollections.php:82]: "document" is not a native php type. The term should be defined somewhere. Maybe update the message in `MongoDB\Exception\InvalidArgumentException`.

```diff
- Expected %s to have type "document" but found "string"
+ Expected %s to have type "document" (array or object) but found "string"
```

File: src/functions.php
Comment (GromNaN) [reply src/functions.php:199]: Ok, but can a list be a valid value for the commands? Can `0` be used as a field name?

If not, it may help developers if they're notified as soon as possible that the value they've passed is a list of documents, when a single document is expected.

**Review by GromNaN — COMMENTED** [review]
It's great to have factored this recurring code pattern.

File: src/functions.php
Comment (alcaeus) [reply src/functions.php:199]: > Ok, but can a list be a valid value for the commands?

Strictly speaking, a list is not a valid document, even though it has been treated as such in PHPLIB/PHPC.

> Can `0` be used as a field name?

Yes. Arrays are serialised like objects in BSON, but denoted through a special type. Likewise, inserting a structure like `[0 => 'foo', 2 => 'bar']` would always yield an object with `"0"` and `"2"` as keys (since the value isn't a list).

> If not, it may help developers if they're notified as soon as possible that the value they've passed is a list of documents, when a single document is expected.

I agree, which is why we're prohibiting `PackedArray`, as we're always sure that it's an array, not an object. For PHP arrays however this isn't as simple and sometimes performance sensitive, which is why it wasn't added before (I'll leave it to @jmikola to confirm this). There is one more difference: with a structure like `[0 => 'foo', 1 => 'bar']`, the trained eye sees an array, but it can easily be used as an object. A `PackedArray` is by definition always an array, which is why we're prohibiting it.

File: src/Command/ListCollections.php
Comment (alcaeus) [reply src/Command/ListCollections.php:82]: Ah, yes. I was trying to avoid "array or object" since that would include all objects including `PackedArray`. I can add it if we want to glance over the packed array case. What do you think @jmikola?

Comment (jmikola) [reply src/Command/ListCollections.php:82]: I thought I left a comment about my concern with using "document" as a type since we use actual PHP types everywhere else. I can't find that so perhaps I deleted the draft before submitting my last review.

In any event, I agree that something like `"document" (array or object)` would be preferable. We don't need to call out PackedArray specifically, as anyone using that should understand that it's not a document -- and if it leads to a bug report it'd be a nice opportunity to educate someone :)

Comment (jmikola) [reply src/Command/ListCollections.php:82]: If you pass `document (array or object)` to the `invalidType()` factory method I suppose it's going to be wrapped in double quotes. Perhaps you just want to pass `document` and let the factory method deal with formatting -- but if that's a bit hairy (especially with the existing logic to handle arrays) then maybe just let it get quoted as-is.

Comment (GromNaN) [reply src/Command/ListCollections.php:82]: > Perhaps you just want to pass document and let the factory method deal with formatting

**Review by jmikola — APPROVED** [review]
LGTM once you add "array or object" to the "document" type exception message.

File: tests/TestCase.php
Comment (alcaeus) [reply tests/TestCase.php]: Yes, static data providers have been allowed for quite some time but have only become required as of PHPUnit 10. We could make that change already.

File: src/Command/ListCollections.php
Comment (alcaeus) [reply src/Command/ListCollections.php]: I've created a new `expectedDocumentType` factory for this case - it seemed a more sensible choice than messing around with the formatting.

**Pull request merged**
